src/sentry/testutils/silo.py

```diff
@@ -92,34 +92,51 @@ def _add_siloed_test_classes_to_module(
-            def run(
-                self: Any, result: unittest.result.TestResult | None = None
-            ) -> unittest.result.TestResult | None:
-                with override_settings(
-                    SILO_MODE=self.silo_mode,
-                    SINGLE_SERVER_SILO_MODE=self.is_acceptance_test,
-                    SENTRY_SUBNET_SECRET="secret",
-                    SENTRY_CONTROL_ADDRESS="http://controlserver/",
-                ):
-                    with override_regions(self.regions):
-                        if self.silo_mode == SiloMode.REGION:
-                            with override_settings(SENTRY_REGION=self.regions[0].name):
-                                return super(test_class, self).run()  # type: ignore
+            final_regions = tuple(regions or _DEFAULT_TEST_REGIONS)
+
+            def decorate_with_context(callable: Callable[[...], Any]) -> Callable[[...], Any]:
+                def wrapper(*args, **kwds):
+                    with contextlib.ExitStack() as stack:
+                        stack.enter_context(
+                            override_settings(
+                                SILO_MODE=silo_mode,
+                                SINGLE_SERVER_SILO_MODE=is_acceptance_test,
+                                SENTRY_SUBNET_SECRET="secret",
+                                SENTRY_CONTROL_ADDRESS="http://controlserver/",
+                            )
+                        )
+                        stack.enter_context(override_regions(final_regions))
+                        if silo_mode == SiloMode.REGION:
+                            stack.enter_context(
+                                override_settings(SENTRY_REGION=final_regions[0].name)
+                            )
                         else:
-                            with override_settings(SENTRY_MONOLITH_REGION=self.regions[0].name):
-                                return super(test_class, self).run()  # type: ignore
+                            stack.enter_context(
+                                override_settings(SENTRY_MONOLITH_REGION=final_regions[0].name)
+                            )
+
+                        return callable(*args, **kwds)
+
+                functools.update_wrapper(wrapper, callable)
+                return wrapper
 
-            return type(
+            new_type = type(
                 name,
                 (test_class,),
                 {
                     "silo_mode": silo_mode,
                     "regions": tuple(regions or _DEFAULT_TEST_REGIONS),
                     "is_acceptance_test": is_acceptance_test,
-                    "run": run,
+                    "_callSetUp": decorate_with_context(test_class._callSetUp),
+                    "_callTestMethod": decorate_with_context(test_class._callTestMethod),
                 },
             )
 
+            if hasattr(test_class, "_overridden_settings") and test_class._overridden_settings:
```

Comment:
What sets `_overridden_settings`? I don't see any references to it in master, and this is the only usage in this diff too.

Reply:
It's how `override_settings` works when applied to classes.  Actually, this isn't the bug, I've been digging and I think it's even more subtle....